overlord/snapstate/snapmgr.go

```diff
@@ -540,6 +546,99 @@ func (m *SnapManager) ensureForceDevmodeDropsDevmodeFromState() error {
 	return nil
 }
 
+// changeInFlight returns true if there is any change in the state
+// in non-ready state.
+func changeInFlight(st *state.State) bool {
+	for _, chg := range st.Changes() {
+		if !chg.Status().Ready() {
+			// another change already in motion
+			return true
+		}
+	}
+	return false
+}
+
+// ensureSnapdSnapTransition will migrate systems to use the "snapd" snap
+func (m *SnapManager) ensureSnapdSnapTransition() error {
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	// check if snapd snap is installed
+	var snapst SnapState
+	err := Get(m.state, "snapd", &snapst)
+	if err != nil && err != state.ErrNoState {
+		return err
+	}
+	// nothing to do
+	if snapst.IsInstalled() {
+		return nil
+	}
+
+	// check if the user opts into the snapd snap
+	optedIntoSnapdTransition, err := optedIntoSnapdSnap(m.state)
+	if err != nil {
+		return err
+	}
+	// nothing to do: the user does not want the snapd snap yet
+	if !optedIntoSnapdTransition {
+		return nil
+	}
+
+	// ensure we only transition systems that have snaps already
+	installedSnaps, err := NumSnaps(m.state)
+	if err != nil {
+		return err
+	}
+	// no installed snaps (yet): do nothing (fresh classic install)
+	if installedSnaps == 0 {
+		return nil
+	}
+
+	// get current core snap and use same channel/user for the snapd snap
+	err = Get(m.state, "core", &snapst)
+	if err != nil && err != state.ErrNoState {
```

Comment:
we should error also on ErrNoState no?

Reply:
can you add a comment then though because otherwise somebody will make the same comment as me at point, something about if there's is no core fix the system installing snapd

Reply:
the point about a comment was not addressed afaict